File: framework/languages/java/e2e/RegistryMessaging/Server/Provider/src/main/java/systems/zlink/e2e/registrymessaging/provider/Endpoints/ProviderEndpoints.java
package systems.zlink.e2e.registrymessaging.provider.Endpoints;

import java.time.Duration;
import java.util.List;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import systems.zlink.contracts.core.RoutingId;
import systems.zlink.e2e.registrymessaging.provider.Infrastructure.ScenarioState;
import systems.zlink.e2e.registrymessaging.shared.Contracts;
import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.framework.channels.ZLinkRouteClient;

@RestController
public final class ProviderEndpoints {
    private final ScenarioState state;
    private final ZLinkClient client;
    private final ZLinkRouteClient routes;

    public ProviderEndpoints(
        ScenarioState state,
        ZLinkClient client,
        ZLinkRouteClient routes) {
        this.state = state;
        this.client = client;
        this.routes = routes;
    }

    @GetMapping("/health")
    public java.util.Map<String, String> health() {
        return java.util.Map.of("status", "ready", "rid", state.providerRid());
    }

    @GetMapping("/evidence")
    public List<String> evidence() {
        return state.lines();
    }

    @PostMapping("/evidence/clear")
    public java.util.Map<String, String> clearEvidence() {
        state.clear();
        return java.util.Map.of("status", "cleared");
    }

    @PostMapping("/evidence/wait")
    public List<String> waitEvidence(@RequestBody Contracts.EvidenceWaitReq request) {
        long timeout = Math.max(1, Math.min(30000, request.timeoutMilliseconds()));
        return state.waitUntil(line -> line.contains(request.contains()), timeout);
    }

    @PostMapping("/profile/request")
    public Contracts.ProfileRes profileRequest(@RequestBody Contracts.ProfileReq request) {
        return requestProfile(Contracts.API_CHANNEL, request, Duration.ofSeconds(5));
    }

    @PostMapping("/profile/manual")
    public Contracts.ProfileRes profileManual(@RequestBody Contracts.ProfileReq request) {
        return requestProfile("registry.messaging.api.manual", request, Duration.ofSeconds(5));
    }

    @PostMapping("/profile/command")
    public java.util.Map<String, String> profileCommand(@RequestBody Contracts.ProfileMsg command) {
        client.sendToChannel(Contracts.API_CHANNEL, command)
            .packetName("ProfileMsg")
            .submit();
        return java.util.Map.of("status", "sent");
    }

    @PostMapping("/profile/route/request")
    public Contracts.RouteRes routeRequest(@RequestBody Contracts.RouteReq request) {
        return routes.requestToNode(Contracts.ROUTE_CHANNEL, RoutingId.from("api-b"), request)
            .packetName(Contracts.ROUTE_PACKET)
            .timeout(Duration.ofSeconds(5))
            .await(Contracts.RouteRes.class);
    }

    @PostMapping("/profile/route/missing")
    public Contracts.RouteMissingRes routeMissing(@RequestBody Contracts.RouteReq request) {
        try {
            routes.requestToNode(Contracts.ROUTE_CHANNEL, RoutingId.from("missing-rid"), request)
                .packetName(Contracts.ROUTE_PACKET)
                .timeout(Duration.ofMillis(300))
                .await(Contracts.RouteRes.class);
            return new Contracts.RouteMissingRes(false);
        } catch (RuntimeException expected) {
            return new Contracts.RouteMissingRes(true);
        }
    }

    private Contracts.ProfileRes requestProfile(
        String channelName,
        Contracts.ProfileReq request,
        Duration timeout) {
        return client.requestToChannel(channelName, request)
            .packetName("ProfileReq")
            .timeout(timeout)
            .await(Contracts.ProfileRes.class);
    }
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Server/Provider/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/provider/Endpoints/ProviderEndpoints.kt
package systems.zlink.e2e.kotlin.registrymessaging.provider.Endpoints

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.net.URI
import java.time.Duration
import java.util.concurrent.Executors
import org.springframework.context.ConfigurableApplicationContext
import systems.zlink.contracts.core.RoutingId
import systems.zlink.e2e.kotlin.registrymessaging.provider.Configuration.ServerOptions
import systems.zlink.e2e.kotlin.registrymessaging.provider.Infrastructure.EvidenceStore
import systems.zlink.e2e.kotlin.registrymessaging.shared.Contracts
import systems.zlink.e2e.kotlin.registrymessaging.shared.EvidenceWaitReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileMsg
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.RouteMissingRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.ScenarioRoutePingReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ScenarioRoutePingRes
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.framework.channels.ZLinkRouteClient

class ProviderEndpoints(
    private val options: ServerOptions,
    private val context: ConfigurableApplicationContext,
) {
    private val mapper = jacksonObjectMapper()
    private val evidence = context.getBean(EvidenceStore::class.java)
    private val channels = context.getBean(ZLinkClient::class.java)
    private val routes = context.getBean(ZLinkRouteClient::class.java)

    fun start(): HttpServer {
        val uri = URI.create(options.httpUrl)
        val server = HttpServer.create(InetSocketAddress(uri.host, uri.port), 0)
        server.executor = Executors.newCachedThreadPool()
        server.createContext("/health") { exchange ->
            exchange.writeJson(mapOf("status" to "ready", "role" to "provider", "rid" to options.rid))
        }
        server.createContext("/evidence") { exchange -> exchange.writeJson(evidence.snapshot()) }
        server.createContext("/evidence/clear") { exchange ->
            evidence.clear()
            exchange.writeJson(mapOf("status" to "cleared"))
        }
        server.createContext("/evidence/wait") { exchange ->
            val request = exchange.readJson<EvidenceWaitReq>()
            exchange.writeJson(
                evidence.waitUntil(
                    request.contains,
                    Duration.ofMillis(request.timeoutMilliseconds.coerceIn(1, 30000).toLong()),
                ),
            )
        }
        server.createContext("/profile/request") { exchange ->
            val request = exchange.readJson<ProfileReq>()
            exchange.writeJson(requestProfile(Contracts.PROFILE_CHANNEL, request, Duration.ofSeconds(5)))
        }
        server.createContext("/profile/manual") { exchange ->
            val request = exchange.readJson<ProfileReq>()
            exchange.writeJson(requestProfile(Contracts.PROFILE_MANUAL_CHANNEL, request, Duration.ofSeconds(5)))
        }
        server.createContext("/profile/command") { exchange ->
            val command = exchange.readJson<ProfileMsg>()
            sendProfile(Contracts.PROFILE_CHANNEL, command, Contracts.PROFILE_COMMAND_PACKET)
            exchange.writeJson(mapOf("status" to "sent"))
        }
        server.createContext("/profile/route/request") { exchange ->
            val request = exchange.readJson<ScenarioRoutePingReq>()
            exchange.writeJson(requestRoute(RoutingId.from("api-b"), request))
        }
        server.createContext("/profile/route/missing") { exchange ->
            val request = exchange.readJson<ScenarioRoutePingReq>()
            var failed = false
            try {
                routes.requestTo(Contracts.PROFILE_ROUTE_CHANNEL, RoutingId.from("missing-rid"), request)
                    .packetName(Contracts.ROUTE_PACKET)
                    .timeout(Duration.ofMillis(300))
                    .await(ScenarioRoutePingRes::class.java)
            } catch (_: RuntimeException) {
                failed = true
            }
            exchange.writeJson(RouteMissingRes(failed))
        }
        server.createContext("/shutdown") { exchange ->
            exchange.writeJson(mapOf("status" to "stopping"))
            Thread {
                server.stop(0)
                context.close()
            }.start()
        }
        server.start()
        return server
    }

    private fun requestProfile(channelName: String, request: ProfileReq, timeout: Duration): ProfileRes {
        val deadline = System.nanoTime() + Duration.ofSeconds(30).toNanos()
        var last: RuntimeException? = null
        while (System.nanoTime() < deadline) {
            try {
                return channels.requestToChannel(channelName, request)
                    .packetName(Contracts.PROFILE_REQUEST_PACKET)
                    .timeout(timeout)
                    .await(ProfileRes::class.java)
            } catch (error: RuntimeException) {
                last = error
                Thread.sleep(100)
            }
        }
        throw IllegalStateException("Timed out waiting for profile request channel route.", last)
    }

    private fun sendProfile(channelName: String, command: ProfileMsg, packetName: String) {
        channels.sendToChannel(channelName, command)
            .packetName(packetName)
            .submit()
    }

    private fun requestRoute(target: RoutingId, request: ScenarioRoutePingReq): ScenarioRoutePingRes {
        val deadline = System.nanoTime() + Duration.ofSeconds(30).toNanos()
        var last: RuntimeException? = null
        while (System.nanoTime() < deadline) {
            try {
                return routes.requestTo(Contracts.PROFILE_ROUTE_CHANNEL, target, request)
                    .packetName(Contracts.ROUTE_PACKET)
                    .timeout(Duration.ofSeconds(5))
                    .await(ScenarioRoutePingRes::class.java)
            } catch (error: RuntimeException) {
                last = error
                Thread.sleep(100)
            }
        }
        throw IllegalStateException("Timed out waiting for route mesh target.", last)
    }

    private inline fun <reified T> HttpExchange.readJson(): T =
        requestBody.use { mapper.readValue(it) }

    private fun HttpExchange.writeJson(value: Any) {
        try {
            val bytes = mapper.writeValueAsBytes(value)
            responseHeaders.add("content-type", "application/json")
            sendResponseHeaders(200, bytes.size.toLong())
            responseBody.use { it.write(bytes) }
        } catch (error: Exception) {
            val bytes = mapper.writeValueAsBytes(mapOf("error" to (error.message ?: error.javaClass.name)))
            sendResponseHeaders(500, bytes.size.toLong())
            responseBody.use { it.write(bytes) }
        }
    }
}
